docs/adr/007_cups_firmware_upgrade.md

```diff
@@ -0,0 +1,64 @@
+# 007. CUPS Protocol Implementation - Firmware Upgrade
+
+**Feature**:
+[#1189](https://github.com/Azure/iotedge-lorawan-starterkit/issues/1189)  
+
+**Authors**: Daniele Antonio Maggio, Maggie Salak
+
+**Status**: Proposed
+
+## Overview
+
+Firmware upgrades for LoRa Basics™ Station need to be supported in the CUPS
+protocol implementation, since certain parts of the information exchanged
+between the Basics Station and the CUPS server are tied to a number of
+fields exchanged between the Station and the CUPS server must correspond to the
+current version of the Station.
+
+More information on the protocol and can be found in 
+[The CUPS protocol - documentation][cupsproto].
+
+## In-scope
+
+This document focuses on:
+
+- Defining a flow for executing firmware upgrades of the Basics Station
+- Defining the changes needed in LoRaWan Network Server (LNS) for handling firmware
+  upgrades
+- Defining the changes needed in device twins for the concentrator station
+  stored in IoT Hub for handling firmware upgrades
+- Defining the changes required in the storage solution used for the CUPS
+  protocol implementation
+- Defining the changes needed in LoRa Device Provisioning CLI for allowing
+  firmware upgrades
+
+## Out-of-scope
+
+## Decisions
+
+### Context
+
+The CUPS request described in [The CUPS protocol - documentation][cupsproto]
+contains a number of fields which are related to the current firmware version of
+the Basics Station. Specifically, `station`, `model`, `package` and `keys`
+fields are dependent on the Station version and need to be updated whenever a
+firmware upgrade of the Basics Station is performed.
+
+### Firmware upgrade flow
```

Comment:
We need a good understanding of the flow for the signing keys/signatures and how those are accessed when ingesting. I would tend to get them directly on the command line when ingesting a new firmware - alternatively, key vault could be supported.

Reply:
From my point of view the flow should be the following:

- When provisioning a device that will need firmware upgrades the end user should generate and store on the device and in a centralized repository of his own choice a "sig-0.key" (example on how to generate it [here](https://github.com/lorabasics/basicstation/blob/master/examples/cups/prep.sh))
- When a firmware upgrade should be provided to the device, user should generate a digest of the executable file of the upgrade (as per example above) and retrieve the CRC32 Checksum of the sig-0.key. Then 4 inputs (file, digest, checksum and new version) should be provided to Device Provisioning CLI tool to properly upload the file to a blob storage and store in the concentrator twin the needed information (blob url in storage, digest, checksum and the new version #)

Since this is a starter kit, I would not force users to use KeyVault for storing and retrieving the sig-*.key files, even though we can definitely suggest it in the docs.

What do you think about it @p-schuler, @maggiesalak?

Reply:
Makes sense, I update the flow to reflect the steps you mentioned - generating a key and then using the CLI with correct inputs to initiate the upgrade.